File: src/texture.rs
use image::GenericImageView;
use std::path::Path;

pub struct Texture {
    pub width: u32,
    pub height: u32,
    pub data: Vec<u8>,
}

impl Texture {
    pub fn new(width: u32, height: u32) -> Self {
        let size = (width * height * 3) as usize;
        Texture {
            width,
            height,
            data: vec![127; size],
        }
    }

    pub fn from_image(path: &Path) -> Result<Self, image::ImageError> {
        let img = image::open(path)?;
        let (width, height) = img.dimensions();
        let rgb_img = img.to_rgb8();

        let data: Vec<u8> = rgb_img.into_raw();

        Ok(Texture {
            width,
            height,
            data,
        })
    }

    pub fn create_flat_normal_map() -> Self {
        let mut texture = Texture {
            width: 1,
            height: 1,
            data: vec![0; 3],
        };
        texture.data[0] = 128;
        texture.data[1] = 128;
        texture.data[2] = 255;
        texture
    }
}

File: src/texture.cpp
#include "texture.hpp"

#define STB_IMAGE_IMPLEMENTATION
#include <stb_image.h>

#include <iostream>

Texture Texture::createFlatNormalMap() {
    Texture tex;
    tex.width = 1;
    tex.height = 1;
    tex.data.resize(4);
    tex.data[0] = 127;  // R
    tex.data[1] = 127;  // G
    tex.data[2] = 255;  // B
    tex.data[3] = 255;  // A
    return tex;
}

Texture Texture::fromFile(const std::string& path) {
    int width, height, channels;
    unsigned char* data = stbi_load(path.c_str(), &width, &height, &channels, 4);
    
    if (!data) {
        std::cerr << "Warning: Failed to load texture (" << path << "), using gray" << std::endl;
        return Texture();
    }
    
    Texture tex;
    tex.width = static_cast<uint32_t>(width);
    tex.height = static_cast<uint32_t>(height);
    tex.data.assign(data, data + width * height * 4);
    
    stbi_image_free(data);
    return tex;
}
